src/app/server-cluster/ServerClusterInterface.h

```diff
@@ -0,0 +1,198 @@
+    ///   - ClusterRevision::Id - this is implementation defined
+    ///
+    /// This DOES NOT need to handle attributes built out of cluster metadata. No need to handle:
+    ///     - AcceptedCommandList::Id
+    ///     - AttributeList::Id
+    ///     - GeneratedCommandList::Id
+    ///
+    virtual DataModel::ActionReturnStatus ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                        AttributeValueEncoder & encoder) = 0;
+
+    /// WriteAttribute MUST be done on a valid attribute path. `request.path` is expected to have `GetClusterId` as the cluster
+    /// id as well as an attribute that is included in a `Attributes` call.
+    ///
+    /// Must only be implemented if writable attributes are defined.
+    ///
+    /// Default implementation errors out with an unsupported write on every attribute.
+    virtual DataModel::ActionReturnStatus WriteAttribute(const DataModel::WriteAttributeRequest & request,
+                                                         AttributeValueDecoder & decoder);
+
+    /// Attribute list MUST contain global attributes.
+    ///
+    /// Specifically these attributes MUST always exist in the list for all clusters:
+    ///     - ClusterRevision::Id
+    ///     - FeatureMap::Id
+    ///     - AcceptedCommandList::Id
+    ///     - AttributeList::Id
+    ///     - GeneratedCommandList::Id
+    /// See SPEC 7.13 Global Elements: `Global Attributes` table
+    ///
+    /// Must only be implemented if support for any non-global attributes
+    /// is required.
+    ///
+    /// Default implementation just returns the above global attributes.
+    virtual CHIP_ERROR Attributes(const ConcreteClusterPath & path, DataModel::ListBuilder<DataModel::AttributeEntry> & builder);
+
+    ///////////////////////////////////// Command Support /////////////////////////////////////////////////////////
+
+    /// Must only be implemented if commands are supported by the cluster
+    ///
+    /// Default implementation errors out with an UnspportedCommand error.
+    virtual std::optional<DataModel::ActionReturnStatus> Invoke(const DataModel::InvokeRequest & request,
+                                                                chip::TLV::TLVReader & input_arguments, CommandHandler * handler);
+
+    /// Must only be implemented if commands are supported by the cluster
+    ///
+    /// Default implementation is a NOOP (no list items generated)
+    virtual CHIP_ERROR AcceptedCommands(const ConcreteClusterPath & path,
+                                        DataModel::ListBuilder<DataModel::AcceptedCommandEntry> & builder);
+
+    /// Must only be implemented if commands are supported by the cluster that return values
+    ///
+    /// Default implementation is a NOOP (no list items generated)
+    virtual CHIP_ERROR GeneratedCommands(const ConcreteClusterPath & path, DataModel::ListBuilder<CommandId> & builder);
+
+    // Internal details for processing
```

Comment:
If these are internal, could they be part of a separate base class?

Reply:
A bit more difficult since the pointer is `Self` i.e. `ServerClusterInterface *`.

I could make it a template base class, however not sure if that helps or hurts readability. Thoughts?

Reply:
Yes. I think the bottom line may be that "RegisterableClusterServer" desccends from ClusterServerInterface, and the registration logic is what is hidden in the Registerable part, but the Data Model Provider onlyc alls ClusterServerInterface?

Reply:
Done with the split, placed the template in `Detail`. A bit of casting going on, hoping it is ok.